pkg/liquidity-mining/contracts/GaugeWorkingBalanceHelper.sol

```diff
@@ -0,0 +1,153 @@
+     * the gauge), and adding a "boost" proportional to that user's share of the total voting power.
+     *
+     * @param gauge - address of a gauge (L1 or L2).
+     * @param user - address of a user.
+     * @return current and projected balances.
+     */
+    function getWorkingBalances(IGauge gauge, address user) public view returns (uint256, uint256) {
+        uint256 gaugeUserBalance = gauge.balanceOf(user);
+        uint256 projectedWorkingBalance = gaugeUserBalance.mulDown(_TOKENLESS_PRODUCTION);
+        IVeDelegationProxy proxy = _veDelegationProxy;
+
+        uint256 veTotalSupply = _readTotalSupplyFromVE ? _veBAL.totalSupply() : proxy.totalSupply();
+
+        if (veTotalSupply > 0) {
+            uint256 veUserBalance = proxy.adjusted_balance_of(user);
+            uint256 gaugeTotalSupply = gauge.totalSupply();
+
+            projectedWorkingBalance = projectedWorkingBalance.add(
+                gaugeTotalSupply.mulDown(veUserBalance).mulDown(_TOKENLESS_PRODUCTION.complement()).divDown(
+                    veTotalSupply
+                )
+            );
+
+            projectedWorkingBalance = Math.min(gaugeUserBalance, projectedWorkingBalance);
+        }
+
+        return (gauge.working_balances(user), projectedWorkingBalance);
+    }
+
+    /**
+     * @dev There is also a "working" supply, needed to ensure that all the emissions are allocated.
+     * Compute and return the balance/supply ratios. This captures the behavior of other users, and more
+     * accurately reflects this user's relative position.
+     *
+     * @param gauge - address of a gauge (L1 or L2).
+     * @param user - address of a user.
+     * @return ratio of the current `working_balance` of the user to the current `working_supply` of the gauge.
+     * @return ratio of the projected `working_balance` of the user (after `user_checkpoint`),
+     *         to the projected `working_supply` of the gauge.
+     */
+    function getWorkingBalanceToSupplyRatios(IGauge gauge, address user) external view returns (uint256, uint256) {
+        (uint256 currentWorkingBalance, uint256 projectedWorkingBalance) = getWorkingBalances(gauge, user);
+        uint256 currentWorkingSupply = gauge.working_supply();
+
+        if (projectedWorkingBalance == currentWorkingBalance) {
+            // If the balances are equal, the supply values will also be equal, so just compute the ratio once.
+            uint256 ratio = currentWorkingBalance.divDown(currentWorkingSupply);
+
+            return (ratio, ratio);
+        } else {
+            uint256 projectedWorkingSupply = currentWorkingSupply.add(
+                projectedWorkingBalance.sub(currentWorkingBalance)
+            );
+
+            return (
+                currentWorkingBalance.divDown(currentWorkingSupply),
+                projectedWorkingBalance.divDown(projectedWorkingSupply)
+            );
+        }
```

Comment:
```suggestion
        uint256 projectedWorkingSupply = currentWorkingSupply.add(
            projectedWorkingBalance.sub(currentWorkingBalance)
        );

        return (
            currentWorkingBalance.divDown(currentWorkingSupply),
            projectedWorkingBalance.divDown(projectedWorkingSupply)
        );
```

The otpimization is not incorrect, but we hardly care about this small gas saving, specially considering this will be queried offchain. I'd rather keep the simpler impl that always works.